src/windows/views/emojis_listview.py

```diff
@@ -109,48 +106,49 @@ def add_file(self, filepath):
             # Log exception
             log.warning("Failed to import file: {}".format(str(ex)))
 
+    @pyqtSlot(int)
+    def group_changed(self, index=-1):
+        log.info("Group set to: {}".format(index))
+        item = get_app().window.emojiFilterGroup.itemData(index)
+        log.info("Filtering on {}".format(item))
+        self.group_model.setFilterFixedString(item)
+        self.group_model.setFilterKeyColumn(1)
+
+        self.refresh_view()
+
     @pyqtSlot(str)
-    def filter_changed(self):
+    def filter_changed(self, filter_text=None):
+        """Filter emoji with proxy class"""
+        log.info("Filter text: {}".format(filter_text or ""))
+        self.model.setFilterRegExp(QRegExp(filter_text, Qt.CaseInsensitive))
+        self.model.setFilterKeyColumn(0)
         self.refresh_view()
 
-        # Save current emoji filter to settings
-        s = get_settings()
-        setting_emoji_group = s.get('emoji_group_filter') or 'smileys-emotion'
-        current_emoji_group = get_app().window.emojiFilterGroup.currentData()
-        if setting_emoji_group != current_emoji_group:
-            s.set('emoji_group_filter', current_emoji_group)
-
-    def refresh_view(self, *args):
-        """Filter transitions with proxy class"""
-        filter_text = self.win.emojisFilter.text().strip().replace(' ', '.*')
-        current_emoji_group = get_app().window.emojiFilterGroup.currentData()
-        if current_emoji_group != "Show All":
-            # Create regex which combines category and filter text
-            # It's much faster for Qt to do this regex than use a custom
-            # Python filter function
-            # FYI:    smiley.*alien.*|alien.*smiley.*
-            filter_text = "%s.*%s.*|%s.*%s.*" % (filter_text, current_emoji_group, current_emoji_group, filter_text)
-        self.emojis_model.proxy_model.setFilterRegExp(QRegExp(filter_text.replace(' ', '.*')))
-        self.emojis_model.proxy_model.setFilterCaseSensitivity(Qt.CaseInsensitive)
-        self.emojis_model.proxy_model.sort(Qt.AscendingOrder)
+    def refresh_view(self):
+        col = self.model.sortColumn()
+        self.model.sort(col)
 
     def __init__(self, model):
         # Invoke parent init
         QListView.__init__(self)
 
-        # Get a reference to the window object
-        self.win = get_app().window
+        # Get external references
+        app = get_app()
+        _ = app._tr
+        self.win = app.window
 
         # Get Model data
         self.emojis_model = model
+        self.group_model = self.emojis_model.group_model
+        self.model = self.emojis_model.proxy_model
```

Comment:
Because we need to access the `group_model` for the filtering, I kept references to all of the models here. But under normal circumstances, the view should only have to concern itself with the final proxy model in the chain (the one I labeled `self.model` here, and changed all references to use `self.model` except the one that needs `self.group_model`. `self.emojis_model` is actually never directly referenced again, once the proxy models are connected up.)

If not for the group filtering, it'd be possible to instantiate the view class with the final proxy model, _instead of_ the source model, and have it only use that. Which is probably what we should be doing, and I should've moved the dropdown list code to `main_window.py` instead, since that's where the list lives. Then we wouldn't need so much ugly referencing of `self.win.emojiFilterGroup` here in the view. Oh well.

Reply:
(A better design might be to have a method in `main_window` connected to the `emojiFilterGroup`'s `currentIndexChanged` signal, and have that function update the proxy model directly, relying on signals to update the corresponding view.

...It's tricky because proxying of a model is neither a view function nor a model function — which is kind of entirely the point, it's why proxies exist. Different views may very well share some proxy models, but a particular view may also bring along its own unique proxy models to interpolate the data in a certain way that only makes sense for that view. As long as the underlying _source_ model is persistent, that's not such a bad thing.

In the old Qt 4.6 [proxy model documentation](https://doc.qt.io/archives/4.6/model-view-proxy-models.html), Qt even spelled it out as such, in a direct, concise way that their current docs, while very comprehensive, sadly lack:
> Although it seems appropriate to perform sorting and filtering operations as internal functions of views, this approach does not allow multiple views to share the results of such potentially costly operations. The alternative approach, involving sorting within the model itself, leads to the similar problem where each view has to display items of data that are organized according to the most recent processing operation.
>
> To solve this problem, the model/view framework uses proxy models to manage the information supplied between individual models and views. **Proxy models are components that behave like ordinary models from the perspective of a view, and access data from source models on behalf of that view.** The signals and slots used by the model/view framework ensure that each view is updated appropriately no matter how many proxy models are placed between itself and the source model.